src/rpp/rpp/subjects/replay_subject.hpp

```diff
@@ -0,0 +1,222 @@
+//                   ReactivePlusPlus library
+//
+//           Copyright Aleksey Loginov 2022 - present.
+//  Distributed under the Boost Software License, Version 1.0.
+//     (See accompanying file LICENSE_1_0.txt or copy at
+//           https://www.boost.org/LICENSE_1_0.txt)
+//
+//  Project home: https://github.com/victimsnino/ReactivePlusPlus
+
+#pragma once
+
+#include <rpp/schedulers/fwd.hpp>
+#include <rpp/subjects/fwd.hpp>
+
+#include <rpp/observers/observer.hpp>
+#include <rpp/subjects/details/base_subject.hpp>
+#include <rpp/subjects/details/subject_state.hpp>
+
+#include <list>
+#include <optional>
+#include <utility>
+
+namespace rpp::subjects::details
+{
+template<rpp::constraint::decayed_type Type, bool Serialized>
+class replay_strategy
+{
+    struct replay_state final : public subject_state<Type>
+    {
+        replay_state(std::optional<size_t> count, std::optional<rpp::schedulers::duration> duration)
+            : count(count)
+            , duration(duration)
+        {
+        }
+
+        auto collect_duration()
+        {
+            if (duration.has_value())
+            {
+                auto now = rpp::schedulers::clock_type::now();
+                while (!values.empty() && (now - values.front().second > duration.value()))
+                {
+                    values.pop_front();
+                }
+                return now;
+            }
+            return rpp::schedulers::clock_type::time_point{};
+        }
+
+        void collect_bound()
+        {
+            if (count.has_value())
+            {
+                if (values.size() == count.value())
+                {
+                    values.pop_front();
+                }
+            }
+        }
+
+        template<typename T>
+        void collect(T&& v)
+        {
+            std::unique_lock lock{list_mutex};
+            collect_bound();
+            const auto time_point = collect_duration();
+
+            values.emplace_back(std::forward<T>(v), time_point);
+        }
+
+        std::optional<size_t>                    count;
+        std::optional<rpp::schedulers::duration> duration;
+
+        std::list<std::pair<Type, rpp::schedulers::clock_type::time_point>> values{};
+
+        std::mutex list_mutex{};
+        std::mutex serialized_mutex{};
+    };
+
+    struct observer_strategy
+    {
+        std::shared_ptr<replay_state> state;
+
+        template<typename T>
+        void collect_and_on_next(T&& v) const
+            requires Serialized
+        {
+            state->collect(std::forward<T>(v));
+
+            std::unique_lock lock{state->serialized_mutex};
+            state->on_next(v);
+        }
+
+        template<typename T>
+        void collect_and_on_next(T&& v) const
+        {
+            state->collect(std::forward<T>(v));
+            state->on_next(v);
```

Comment:
actually it's an bug. What if you've moved value during `collect`? You need to send .back() then instead

Reply:
done